### PR TITLE
sql: new built-in generator pg_get_keywords

### PR DESCRIPTION
**Pull request by knz**
Requested in https://github.com/cockroachdb/cockroach/issues/10291#issuecomment-334985732.

This PR includes some other useful refactors and cleanups.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/generators [hunk 1]
```diff
@@ -242,3 +242,12 @@ SELECT GENERATE_SERIES(9223372036854775807::int, -9223372036854775807::int, -922
 9223372036854775807
 0
 -9223372036854775807
+
+# pg_get_keywords for compatibility (#10291)
+query TTT
```
Comment (justinj) [inline pkg/sql/logictest/testdata/logic_test/generators]: Probably worth doing a `colnames` for this test?

File: pkg/sql/generator.go [hunk 2]
```diff
@@ -62,17 +58,10 @@ func (p *planner) makeGenerator(ctx context.Context, t *parser.FuncExpr) (planNo
...
-	} else {
-		columns = make(sqlbase.ResultColumns, len(tType.Cols))
-		for i, t := range tType.Cols {
-			columns[i] = sqlbase.ResultColumn{
-				Name: fmt.Sprintf("column%d", i+1),
-				Typ:  t,
-			}
-		}
+	columns := make(sqlbase.ResultColumns, len(tType.Cols))
+	for i := range columns {
```
Comment (justinj) [inline pkg/sql/generator.go]: I just realized PG lets you specify the label for a single-column generator:
```
select * from unnest(array[1,2,3]) a;
+--------+
| unnest |
+--------+
|      1 |
|      2 |
|      3 |
+--------+
(3 rows)
```
is that something we can easily support?

**Review by justinj — APPROVED** [review]
LGTM (with two comments), I wonder why PG chose to make this a generator function rather than a system table.

**Comment by knz** [comment]
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/generator.go, line 62 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw0WmB64FGXdS00Itwt-r4-62:-Kw0fmvcX8Gv1ceXvxYS:b-cvks6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9549f26a7df6ae4e5652b3cba3b563f0e1ffbe72/pkg/sql/generator.go#L62)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I just realized PG lets you specify the label for a single-column generator:
```
justin=# select * from unnest(array[1,2,3]) a;
 a
---
 1
 2
 3
(3 rows)
```
is that something we can easily support?
</blockquote></details>

Awwww bad pg, bad. This is a weird mechanism. I'll investigate. However it's separate, it wasn't supported before this PR anyways.

---

*[pkg/sql/logictest/testdata/logic_test/generators, line 247 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw0WmB64FGXdS00Itwx-r4-247:-Kw0fpvjXS8EWVCq7fPY:b-5e84w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9549f26a7df6ae4e5652b3cba3b563f0e1ffbe72/pkg/sql/logictest/testdata/logic_test/generators#L247)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Probably worth doing a `colnames` for this test?
</blockquote></details>

Yes indeed. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127)*

**Comment by knz** [comment]
Added a commit with the pg behavior, PTAL

**Comment by nvanbenschoten** [comment]
Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 6 of 6 files at r5, 4 of 4 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/data_source.go, line 491 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw1Kme5I8jwus4DdWjZ:-Kw1Kme5I8jwus4DdWj_:bvux3ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d69e26ab81f6e8da792059c16bb50f7261a052d/pkg/sql/data_source.go#L491)):*
> ```Go
> 		// column name and table name.
> 		if vg, ok := src.plan.(*valueGenerator); ok &&
> 			(len(src.info.sourceAliases) == 0 ||
> ```

pull out these two conditions into variables

---

*[pkg/sql/parser/generator_builtins.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw19TlP5L6wpg2uaVxp:-Kw19TlP5L6wpg2uaVxq:bk3tiy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd9be0212c0248972a3ed265f26a8fe4101fa2c1/pkg/sql/parser/generator_builtins.go#L237)):*
> ```Go
> // ResolvedType implements the ValueGenerator interface.
> func (s *arrayValueGenerator) ResolvedType() TTable {
> 	return TTable{
> ```

Consider pulling this into a static variable too.

---

*[pkg/sql/parser/generator_builtins.go, line 220 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw1K5dOMb1clTPyNr9R:-Kw1K5dOMb1clTPyNr9S:b-yn75z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa558f72dab927e18df2e1a9325cd14492971078/pkg/sql/parser/generator_builtins.go#L220)):*
> ```Go
> // that pg_get_keywords returns deterministic results.
> var keywordNames = func() []string {
> 	ret := make([]string, len(keywords))
> ```

nit: efficiency doesn't matter here so just set the capacity and append to get rid of `i`.

---

*[pkg/sql/parser/sql.y, line 5617 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw18SCUibknK-UzoHNf:-Kw18SCUibknK-UzoHNg:b-msdhdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e37da0785c9ea74247dd7bd37702a818f52b62a/pkg/sql/parser/sql.y#L5617)):*
> ```Yacc
> | /* EMPTY */
>   {
>     $$.val = Exprs{}
> ```

This is creating a zero-length `Expr` slice. I think you want a `nil` slice. Try `Exprs(nil)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127)*

**Comment by nvanbenschoten** [comment]
:lgtm: mostly just nits. Nice change!

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-:-Kw1LLbG_s1hGOTRB2MJ:b-wg37)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/data_source.go, line 491 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw1Kme5I8jwus4DdWjZ:-Kw1j1ak2XsmO-4w0kVX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d69e26ab81f6e8da792059c16bb50f7261a052d/pkg/sql/data_source.go#L491)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

pull out these two conditions into variables
</blockquote></details>

Done.

---

*[pkg/sql/parser/generator_builtins.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw19TlP5L6wpg2uaVxp:-Kw1jd8Xs3QRq2xuKTnv:bmrv9fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd9be0212c0248972a3ed265f26a8fe4101fa2c1/pkg/sql/parser/generator_builtins.go#L237)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider pulling this into a static variable too.
</blockquote></details>

Can't do, it's dependent on `s`.

---

*[pkg/sql/parser/generator_builtins.go, line 220 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw1K5dOMb1clTPyNr9R:-Kw1kvR_No4do900jruS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa558f72dab927e18df2e1a9325cd14492971078/pkg/sql/parser/generator_builtins.go#L220)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: efficiency doesn't matter here so just set the capacity and append to get rid of `i`.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 5617 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127#-Kw18SCUibknK-UzoHNf:-Kw1jSc2vTkf9kpsVQ2s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e37da0785c9ea74247dd7bd37702a818f52b62a/pkg/sql/parser/sql.y#L5617)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is creating a zero-length `Expr` slice. I think you want a `nil` slice. Try `Exprs(nil)`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19127)*

**Pull request merged**